### PR TITLE
Compute V2: implement remoteconsoles Create method

### PR DESCRIPTION
**Pull request by ozerovandrei**
Add the new "remoteconsoles" Compute extension package with needed
structures and Create method.
Add unit and acceptance test with documentation.

For #1158

Links to the line numbers/files in the OpenStack source code that support the
code in this PR:

API create request handler:
https://github.com/openstack/nova/blob/stable/queens/nova/api/openstack/compute/remote_consoles.py#L160

API definitions:
https://github.com/openstack/nova/blob/stable/queens/nova/api/openstack/compute/schemas/remote_consoles.py#L91
https://github.com/openstack/nova/blob/stable/queens/nova/api/openstack/compute/schemas/remote_consoles.py#L115

Allowed protocols and types:
https://github.com/openstack/nova/blob/stable/queens/nova/api/openstack/compute/schemas/remote_consoles.py#L123
https://github.com/openstack/nova/blob/stable/queens/nova/api/openstack/compute/schemas/remote_consoles.py#L127

### CODE REVIEW

**Comment by ozerovandrei** [comment]
Well, this won't work until we pass the `X-OpenStack-Nova-API-Version` header with at least `2.6` version value.

File: openstack/compute/v2/extensions/remoteconsoles/requests.go [hunk 1]
```diff
@@ -0,0 +1,83 @@
...
+	"github.com/gophercloud/gophercloud"
+)
+
+// RemoteConsoleProtocol represents valid remote console protocol.
+// It can be used to create a remote console with one of the pre-defined protocol.
+type RemoteConsoleProtocol string
+
+const (
+	// RemoteConsoleVNCProtocol represents the VNC console protocol.
+	RemoteConsoleVNCProtocol RemoteConsoleProtocol = "vnc"
```
Comment (jtopjian) [inline openstack/compute/v2/extensions/remoteconsoles/requests.go]: How about `ConsoleProtocolVNC`. Same pattern for the rest.

File: openstack/compute/v2/extensions/remoteconsoles/requests.go [hunk 2]
```diff
@@ -0,0 +1,83 @@
...
+	RemoteConsoleMKSProtocol RemoteConsoleProtocol = "mks"
+)
+
+// RemoteConsoleType represents valid remote console type.
+// It can be used to create a remote console with one of the pre-defined type.
+type RemoteConsoleType string
+
+const (
+	// RemoteConsoleNoVNCType represents the VNC console type.
+	RemoteConsoleNoVNCType RemoteConsoleType = "novnc"
```
Comment (jtopjian) [inline openstack/compute/v2/extensions/remoteconsoles/requests.go]: Similar here: `ConsoleTypeNoVNC`. Same pattern for the rest, where the unique part is at the end of the name.

File: openstack/compute/v2/extensions/remoteconsoles/requests.go [hunk 3]
```diff
@@ -0,0 +1,83 @@
...
+
+// CreateOptsBuilder allows to add additional parameters to the Create request.
+type CreateOptsBuilder interface {
+	ToRemoteConsoleCreateMap() (map[string]interface{}, error)
+}
+
+// CreateOpts specifies parameters to the Create request.
+type CreateOpts struct {
+	// Protocol specifies the protocol of a new remote console.
+	Protocol string `json:"protocol" required:"true"`
```
Comment (jtopjian) [inline openstack/compute/v2/extensions/remoteconsoles/requests.go]: Should this be:

```go
Protocol RemoteConsoleProtocol `json:"protocol" reqauired:"true"`
```

File: openstack/compute/v2/extensions/remoteconsoles/requests.go [hunk 4]
```diff
@@ -0,0 +1,83 @@
...
+
+	// Type specifies the type of a new remote console.
+	Type string `json:"type" required:"true"`
```
Comment (jtopjian) [inline openstack/compute/v2/extensions/remoteconsoles/requests.go]: Should this be:

```go
Type RemoteConsoleType `json:"type" required:"true"`
```

File: openstack/compute/v2/extensions/remoteconsoles/requests.go [hunk 5]
```diff
@@ -0,0 +1,83 @@
+package remoteconsoles
+
+import (
+	"github.com/gophercloud/gophercloud"
+)
+
+// RemoteConsoleProtocol represents valid remote console protocol.
+// It can be used to create a remote console with one of the pre-defined protocol.
+type RemoteConsoleProtocol string
```
Comment (jtopjian) [inline openstack/compute/v2/extensions/remoteconsoles/requests.go]: Since the package name is `remoteconsoles`, I think we can shorten this to:

```go
type ConsoleProtocol string
```

And the name will still have a clear description of the type.

File: openstack/compute/v2/extensions/remoteconsoles/requests.go [hunk 6]
```diff
@@ -0,0 +1,83 @@
...
+	// RemoteConsoleSerialProtocol represents the serial console protocol.
+	RemoteConsoleSerialProtocol RemoteConsoleProtocol = "serial"
+
+	// RemoteConsoleMKSProtocol represents the MKS console protocol.
+	RemoteConsoleMKSProtocol RemoteConsoleProtocol = "mks"
+)
+
+// RemoteConsoleType represents valid remote console type.
+// It can be used to create a remote console with one of the pre-defined type.
+type RemoteConsoleType string
```
Comment (jtopjian) [inline openstack/compute/v2/extensions/remoteconsoles/requests.go]: Similar here, we can probably shorten this to:

```go
type ConsoleType string
```

But feel free to keep the original name if you feel it's better.

File: openstack/compute/v2/extensions/remoteconsoles/doc.go [hunk 7]
```diff
@@ -0,0 +1,25 @@
...
+
+Example of Creating a new RemoteConsole
+
+  computeClient, err := openstack.NewComputeV2(providerClient, endpointOptions)
+  computeClient.Microversion = "2.6"
+
+  createOpts := remoteconsoles.CreateOpts{
+    Protocol: string(remoteconsoles.RemoteConsoleVNCProtocol),
+    Type:     string(remoteconsoles.RemoteConsoleNoVNCType),
+	}
```
Comment (jtopjian) [inline openstack/compute/v2/extensions/remoteconsoles/doc.go]: nit: indentation needs fixed here.

File: openstack/compute/v2/extensions/remoteconsoles/requests.go (hunk 3 above)
Comment (ozerovandrei) [reply openstack/compute/v2/extensions/remoteconsoles/requests.go]: It looks like the same issue as we discussed in [PR#1147](https://github.com/gophercloud/gophercloud/pull/1147#issuecomment-407868998).

Do you think that we need to strict users to pre-defined values in case of this change? I don't actually get how it's different from the #1147 😅

**Comment by ozerovandrei** [comment]
> Let's wait for a decision with the other microversion stuff. If it ends up taking too long, I might go ahead and merge this. Thoughts?

I kinda need this API implementation 😅. Also I don't want to create deprecated stuff that can work without microversions (`os-getRDPConsole`, `os-getSerialConsole`, `os-getSPICEConsole`, `os-getVNCConsole`) or to implement my own temporary solution in the project that I work with.

Maybe you can merge this as an exception? This package differs from PRs #1163 or #1115 because microversions won't change some parts of the responses or options, they completely turn it on or off.

File: openstack/compute/v2/extensions/remoteconsoles/requests.go (hunk 1 above)
Comment (ozerovandrei) [reply openstack/compute/v2/extensions/remoteconsoles/requests.go]: Seems better, yeah.

File: openstack/compute/v2/extensions/remoteconsoles/requests.go (hunk 5 above)
Comment (ozerovandrei) [reply openstack/compute/v2/extensions/remoteconsoles/requests.go]: Seems better, yeah.

File: openstack/compute/v2/extensions/remoteconsoles/requests.go (hunk 3 above)
Comment (jtopjian) [reply openstack/compute/v2/extensions/remoteconsoles/requests.go]: No worries - since it's not actually written anywhere, it's hard to grasp :)

The rule is: custom types for "input", standard types for "output". So any type of `CreatOpts`, `UpdateOpts`, etc should have types which can provide better/safer input. Result structures should only contain "standard" (`string`, `int`, `[]string`, etc fields). If a struct contains a struct, then that can be a defined struct.

Comment (ozerovandrei) [reply openstack/compute/v2/extensions/remoteconsoles/requests.go]: Alright, seems legit.

**Comment by ozerovandrei** [comment]
recheck

**Comment by ozerovandrei** [comment]
recheck

**Comment by ozerovandrei** [comment]
recheck

**Comment by ozerovandrei** [comment]
@jtopjian what do you think about current state of this PR after microversion implementation in Gophercloud?

Tricky part is: we can't disable/enable some fields in request opts, response results. All of the Nova API remoteconsoles controller won't be available without setting the Compute microversion.

**Comment by ozerovandrei** [comment]
recheck

**Comment by jtopjian** [comment]
> Tricky part is: we can't disable/enable some fields in request opts, response results. All of the Nova API remoteconsoles controller won't be available without setting the Compute microversion.

If I understand correctly, this should be fine. Well, kind of fine. You will just have to set the microversion on the client when calling making any request to `remoteconsoles`.

**Comment by ozerovandrei** [comment]
recheck
